singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,374 @@
+    return _Cv0;
+  }
+  // Thermal Bulk Modulus added AEM Dec 2022
+  PORTABLE_INLINE_FUNCTION Real TBulkModulusFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real BulkModulusFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  // Thermal expansion coefficient added AEM 2022
+  PORTABLE_INLINE_FUNCTION Real TExpansionCoeffFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const {
+    return _A0 * _B0 / _Cv0 / rho;
+  }
+  PORTABLE_INLINE_FUNCTION Real GruneisenParamFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return _A0 * _B0 / _Cv0 / rho;
+  }
+  PORTABLE_INLINE_FUNCTION void FillEos(Real &rho, Real &temp, Real &energy, Real &press,
+                                        Real &cv, Real &bmod, const unsigned long output,
+                                        Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION
+  void ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                              Real &bmod, Real &dpde, Real &dvdt,
+                              Real *lambda = nullptr) const;
+  // Generic functions provided by the base class. These contain e.g. the vector
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(Vinet)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"Vinet Params: "};
+    printf("%s rho0:%e T0:%e B0:%e BP0:%e\n  A0:%e Cv0:%e E0:%e S0:%e\n", st, _rho0, _T0,
+           _B0, _BP0, _A0, _Cv0, _E0, _S0);
+    for (int i = 0; i < 39; i++) {
+      if (_d2to40[i] > 0.0) printf("d%i:%e\t", i + 2, _d2to40[i]);
+    }
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Vinet"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  // AEM: I am not sure what this does so leave it in.
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::temperature;
+  Real _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0;
+  const Real *_expconsts;
+  Real _VIP[43], _d2to40[39];
```

Comment:
I wanted to add if I didn't say elsewhere, it would be useful to have more descriptive names for d2to40 and VIP or at least comments explaining what these arrays are.

Reply:
@Yurlungur does the resize do a malloc or a `PORTABLE_MALLOC`, also, the frees will need to be `PORTABLE_FREE`s

Reply:
> Doing so will require a few things. You'll need to have the code be only conditionally enabled when spiner is enabled with
> 
> ```c++
> #ifdef SPINER_USE_HDF
> ```
> 
> at the top of the file and
> 
> ```c++
> #endif // SPINER_USE_HDF
> ```
> 
> at the bottom.

It's highly likely this will change soon, or at least be deprecated. @aematts I will tag you on PR(s) that would be relevant here.

Reply:
@Yurlungur if we took the variadic route instead of passing a single array, wouldn't that obviate the need for spiner?

I'm a bit nervous about adding a spiner dependency to an analytic EOS _just_ to store static parameters, but I understand your concerns about host to device transfers of data that could be an arbitrary length.

Reply:
@Yurlungur is this `free` you talk about name-spaced? It is a bit concerning to have a function with the exact same name as a C function in which the symbols are not mangled. Yes yes, ADL will save it and **should** always take precedence in overload resolution but... It scares me. 

In addition, I agree it is a bit sketchy to rely on spiner only for a tangential capability. Using portable array is going to be much more straightforward. Also, c-style VLAs are explicitly unable to be captured into coda device lambdas and std array is not constexpr'd enough to work.

Reply:
> is this free you talk about name-spaced? 

Nope. It's straight up overloaded. You can also use `DataBox::finalize()` if you prefer. But overloading free felt elegant to me. ;P

> Also, c-style VLAs are explicitly unable to be captured into cuda device lambdas

Oops I missed a word above. Meant to say C-style arrays do *not* work.

> std array is not constexpr'd enough to work.

I thought for sure we'd used `std::array` safely in other contexts. Maybe I thought it changed between C++11 and C++14. Maybe I'm mistaken.

Reply:
Regardless I'm not in favor of the `std::array` solution. I think `PORTABLE_MALLOC` or `spiner` are the ways to go.